### PR TITLE
[WIP]MNISTデータセットを使用したDeconvolutionNetwork

### PR DESCRIPTION
**Pull request by YamaneSasuke**
MNISTデータセットを使用し、DeconvolutionNetworkを実装する。

### CODE REVIEW

File: deconvNet_mnist.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+            plt.matshow(y_train[i].reshape(28, 28), cmap=plt.cm.gray)
+            plt.show()
+
+    except KeyboardInterrupt:
+        print "割り込み停止が実行されました"
+
+    # テストデータでの結果を表示
+    y_test = model.predict(X_test_gpu)
+    loss_test = model.lossfun(X_test_gpu, T_test_gpu)
+    print "[test] loss:", loss_test
```
Comment (ronekko) [inline deconvNet_mnist.py]: 損失の値が表示されてません．（ `loss_test.data` にする必要がある．）

File: deconvNet_mnist.py [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+                                                          test_size=0.1,
+                                                          random_state=10)
+    X_test = np.random.permutation(X_test)
+    T_test = np.random.permutation(T_test)
+    X_train = X_train[0:1000]
+    T_train = T_train[0:1000]
+    X_valid = X_valid[0:100]
+    T_valid = T_valid[0:100]
+    X_test = X_test[0:100]
+    T_test = T_test[0:100]
```
Comment (ronekko) [inline deconvNet_mnist.py]: 検証/テストデータの一部しか使われてません．

File: deconvNet_mnist.py [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+            conv1=L.Convolution2D(1, 15, 3, pad=1),
+            conv2=L.Convolution2D(15, 30, 3, pad=1),
+            deconv1=L.Deconvolution2D(30, 15, 3, pad=1),
+            deconv2=L.Deconvolution2D(15, 1, 3, pad=1)
+        )
+
+    def forward(self, X):
+        x = Variable(X.reshape(-1, 1, 28, 28))
+        h = self.conv1(x)
+        h1 = F.relu(h)
```
Comment (ronekko) [inline deconvNet_mnist.py]: `h` と `h1` や `h2` という変数名の使い方に一貫性がないためバグが入り込む可能性があります． `h` で統一したほうがいいと思います．

File: deconvNet_mnist.py [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+        h = F.max_pooling_2d(h1, 2,)
```
Comment (ronekko) [inline deconvNet_mnist.py]: `2` の後ろのコンマが不要です．

File: deconvNet_mnist.py [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+        h = self.conv2(h)
+        h2 = F.relu(h)
+        h = F.max_pooling_2d(h2, 2,)
```
Comment (ronekko) [inline deconvNet_mnist.py]: `2` の後ろのコンマが不要です．

File: deconvNet_mnist.py (hunk 3 above)
Comment (ronekko) [inline deconvNet_mnist.py]: `h1` としてるのはshapeを後で使うからでしょうか？だとしたら `h1_shape` みたいにshapeそのものを変数にいれておいたほうが安全な気がします．

Comment (ronekko) [inline deconvNet_mnist.py]: `h1_shape` より `h1_size = h.data.shape[2:]` のほうがもっといいですね．

File: deconvNet_mnist.py [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+        ys = []
+        total_data = np.arange(len(X))
+        for indexes in np.array_split(total_data, num_batches):
+            X_batch = cuda.to_gpu(X[indexes])
+            T_batch = cuda.to_gpu(T[indexes])
+            y = cuda.to_cpu(self.forward(X_batch, test).data)
+            loss = cuda.to_cpu(self.lossfun(X_batch, T_batch, test).data)
+            losses.append(loss)
+#            ys.append(y)
+        return np.mean(losses), y
```
Comment (ronekko) [inline deconvNet_mnist.py:86]: lossは全サンプルの平均になってますが，yは最後のミニバッチだけしか含まれてません．そのため，main側のX[0]とy[0]がずれてます．

Comment (ronekko) [inline deconvNet_mnist.py:86]: ysはその名残ということですか． `np.concatenate(ys, axis=0)` で全サンプルを結合できました．

File: deconvNet_mnist.py [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+        return loss, y
+
+    def predict(self, X, num_batches, test):
+        ys = []
+        total_data = np.arange(len(X))
+        for indexes in np.array_split(total_data, num_batches):
+            X_batch = cuda.to_gpu(X[indexes])
+            y = cuda.to_cpu(model.forward(X_batch, test).data)
+            ys.append(y)
+        return ys
```
Comment (ronekko) [inline deconvNet_mnist.py:72]: ここも `np.concatenate(ys, axis=0)` 結合できます．

File: deconvNet_mnist.py [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+
+    # テストデータでの結果を表示
+    y_test = model.predict(X_test, num_batches, True)
+#    loss_test = model.lossfun_ave(X_test, T_test, num_batches, True)
+#    print "[test] loss:", loss_test.data
+    print "max_iteration:", max_iteration
+    print "batch_size:", batch_size
+    print "learning_rate", learning_rate
+    print "X_test"
+    draw_filters(X_test, 10)
```
Comment (ronekko) [inline deconvNet_mnist.py:255]: 全サンプルを一枚の画像におさめて表示しようとしているため，画像が潰れてます． `X_test[:100]` みたいにして表示するサンプルの数を少量にする必要があります．

File: deconvNet_mnist.py [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+    print "predict_test"
+    draw_filters(y_test, 10)
```
Comment (ronekko) [inline deconvNet_mnist.py:257]: 同上．
（全サンプルを一枚の画像におさめて表示しようとしているため，画像が潰れてます． `X_test[:100]` みたいにして表示するサンプルの数を少量にする必要があります．）
